src/advanced_unsafety/invalid_values.md

```diff
@@ -0,0 +1,123 @@
+# Invalid values
+
+> _“If you tell the truth, you don't have to remember anything.”__
+>  — _Mark Twain_
+
+Values of a particular type in Rust may never have an "invalid" bit pattern for that type. This is true even if that value is never read from afterwards.
+
+A lot of basic types _don't_ have any rules about invalid values. All bit patterns of the integer types (and arrays of the integer types) are valid.
+
+But most other types have some concept of validity.
+
+## Types of invalid values
+
+### Primitive types with invalid values
+
+`bool`s that have bit patterns other than those for `true` and `false` are invalid. The same goes for `char`s representing byte patterns that are considered invalid in UTF-32.
+
+
+`&T` and `&mut T` may not be null, nor may they be [unaligned] for values of type `T`. There are a lot of other reasons that a reference may not be valid, but these are the ones where the bit pattern is statically known to be invalid regardless of context.
+
+### Enums with invalid values
+
+
+Any bit pattern not covered by a variant of an enum is also invalid. For example, with the following enum:
+
+```rust
+enum Colors {
+    Red = 1,
+    Orange = 2,
+    Yellow = 3,
+    Green = 4,
+    Blue = 5,
+    Indigo = 6,
+    Violet = 7,
+}
+```
+
+a bit pattern of `8` or `0` (assuming that it gets represented as the explicit discriminant integers) is undefined behavior.
+
+Or in this enum:
+
+```rust
+enum Stuff {
+    Char(char),
+    Number(u32),
+}
+```
+
+setting the discriminant bit to something that is not the discriminant of `Char` or `Number` is invalid. Similarly, setting the discriminant bit to that for `Char` but having the value be invalid for a `char` is also invalid.
+
+### Smart pointers and NonNull
+
+Most smart pointer types like `Box<T>` and `Rc<T>` are invalid when null. Library types may achieve the same behavior using the [`NonNull<T>`] pointer type.
+
+It's also currently invalid for `Vec<T>` to have a null pointer for its buffer! `Vec<T>` uses [`NonNull<T>`] internally, and empty vectors use a pointer value equal to the alignment of `T`.
+
+
+
+### `#[repr(Rust)]` isn't stable!
+
+Note that Rust's default representation for types is not stable! What might be a valid bit pattern one day may become invalid later, unless you're only relying on things that are known to be invariant.
+
+### `str`
+
+The string slice type `str` does not actually have any validity constraints: Despite being only for UTF-8 encoded strings, it is valid for `str`s to be in any bit pattern, provided you do not call any methods on the string that are not about directly accessing the memory behind it.
+
+Basically, the UTF-8 validity of `str` is an implicit safety requirement for most of its methods, however it is fine to _hold on to_ an `&str` that points to random bytes.
```

Comment:
This might be a good opportunity to distinguish between _insta-UB_ and _UB-on-use_, since producing invalid values is _insta-UB_ but invalid UTF-8 in `str` is not insta-UB.